### DISCUSSION TITLE
Make Stylet multiple UI threads safe?

### DISCUSSION THREAD

**Discussion by mwdpb** [canton7/Stylet#217] [General]
Background: in Excel VSTO application, the ```TextBox``` in a WPF window will not get keyboard focus unless the window is modal (if you type anything, it will go to the Excel cell that the cursor was in before launching the WPF window, see [https://stackoverflow.com/questions/5869359/wpf-modeless-dialog-from-ms-excel-add-in](url).)  You can get away with it by using ```WindowManager.ShowDialog(new VM())``` which makes it modal.  However, it isn't ideal as that means you can't go back to Excel without closing the current window.  I encountered this exact problem when I was using Caliburn.Micro a few years ago and there is an elegant solution by creating a new UI thread and everything works fine.  The code snippet is here:
![image](https://user-images.githubusercontent.com/55845355/107854863-d2b94a00-6dec-11eb-81f6-eeaaa74d34b0.png)

Granted I encountered the exact problem with Stylet, and tried the same approach and it errors out in ```CommandAction.UpdateCanExecute```, see screen capture below:
![image](https://user-images.githubusercontent.com/55845355/107855014-dac5b980-6ded-11eb-9e48-5095f133ba5b.png)

If I go ahead and comment out that ```if``` statement, it works fine except I don't get command guards anymore.  It seems Stylet assume single UI thread and that ```if``` statement trying to go back to the main UI but since this window was created in a separate UI thread, it couldn't access its objects.  How hard to make Stylet multiple UI thread safe?  Any thoughts?

FWIW, here is an article talking about multiple UI thread application.  We certainly don't need multiple UI threads everyday, but there are situations that do justify creating separate UI thread.
[https://eprystupa.wordpress.com/2008/07/28/running-wpf-application-with-multiple-ui-threads/](url)

**Comment by canton7** [comment]
If you set `Execute.Dispatcher = null` in your bootstrapper's `Configure` method, does that fix the issue? That will make `Execute.OnXXX` methods dispatch synchronously, rather than trying to dispatch to the Application's UI thread. Of course, you're then responsible for making sure you raise events (such as INPC or INCC) on the right thread!

Ultimately though, why not run the entire application on a new thread, rather than juggling two UI threads?

**Reply by mwdpb** [reply]
I'll try your suggestion later.  To your last question, I think it is some particular to Excel VSTO.  I actually started a brand new ```Application``` in a separate thread when the VSTO addin is loaded.  Here is what's inside the VSTO addin startup event:
![image](https://user-images.githubusercontent.com/55845355/107856631-7871b680-6df7-11eb-8741-f57ecbeb7cdf.png)

And here is ```AppHelpers``` properties and ctor:
![image](https://user-images.githubusercontent.com/55845355/107856653-9a6b3900-6df7-11eb-91f7-2d12939dcdd4.png)

As you can see I basically started a brand new ```AppDomain``` and bootstrapped Stylet with the newly created ```Application```.  I was expecting all the subsequent windows will be tied to this thread and honestly I think they do.  However due to the aforementioned Excel particularity, the keyboard event went to Excel's own thread and I have no idea how to solve it without either using modal dialog or creating new UI thread.

**Reply by mwdpb** [reply]
> If you set `Execute.Dispatcher = null` in your bootstrapper's `Configure` method, does that fix the issue? That will make `Execute.OnXXX` methods dispatch synchronously, rather than trying to dispatch to the Application's UI thread. Of course, you're then responsible for making sure you raise events (such as INPC or INCC) on the right thread!
> 
> Ultimately though, why not run the entire application on a new thread, rather than juggling two UI threads?

So I tried setting `Execute.Dispatcher = null` and it seems you are checking its value and throw an exception if it's `null`, see error below:

![image](https://user-images.githubusercontent.com/55845355/107858357-b5db4180-6e01-11eb-9ee4-4b422504c08a.png)

**Reply by canton7** [reply]
Ah. Set it to `new SynchronousDispatcher()` then

**Reply by canton7** [reply]
Ah no, that's internal....

Copy the implementation of [SynchronousDispatcher](https://github.com/canton7/Stylet/blob/master/Stylet/IDispatcher.cs#L54) for now then, and if it works I'll expose it (or allow null)

**Reply by mwdpb** [reply]
> > If you set `Execute.Dispatcher = null` in your bootstrapper's `Configure` method, does that fix the issue? That will make `Execute.OnXXX` methods dispatch synchronously, rather than trying to dispatch to the Application's UI thread. Of course, you're then responsible for making sure you raise events (such as INPC or INCC) on the right thread!
> > Ultimately though, why not run the entire application on a new thread, rather than juggling two UI threads?
> 
> So I tried setting `Execute.Dispatcher = null` and it seems you are checking its value and throw an exception if it's `null`, see error below:
> 
> ![image](https://user-images.githubusercontent.com/55845355/107858357-b5db4180-6e01-11eb-9ee4-4b422504c08a.png)

If I disable that ```null``` checking if statement and set `Execute.Dispatcher = null` in bootstrapper's `Configure`, then as you expected, the window show up properly and can take keyboard inputs, but the INPC and INCC don't work anymore.

**Reply by mwdpb** [reply]
> Ah no, that's internal....
> 
> Copy the implementation of [SynchronousDispatcher](https://github.com/canton7/Stylet/blob/master/Stylet/IDispatcher.cs#L54) for now then, and if it works I'll expose it (or allow null)

That worked, but INCC and INPC are gone.  It's like pick your poison situation, I guess I'll stick to modal window for now until I find time to completely understand your code, :)

**Reply by mwdpb** [reply]
Well, actually I'll take back what I said, the **INCC and INPC do work**!!  I mistyped a variable name in raising INPC and once I corrected it, it worked even `Execute.Dispatcher = null`, amazing!

Looking into the `Dispatcher` property one more time, I found two things worth pointing out: 
1> `SynchronousDispatcher` can be seen as it is used by the `get`.
2> I think the `null` check in the `set` is un-necessary because you are doing `null` coalescing in `get` method anyway.

![image](https://user-images.githubusercontent.com/55845355/107859197-d4433c00-6e05-11eb-8113-abaae97285ce.png)

**Reply by canton7** [reply]
Yeah, the reason it's there is that setting null to a property, and then reading back something which isn't null, is pretty weird behaviour. I think exposing SynchronousDispatcher makes most sense: it's clearer as well.

**Reply by mwdpb** [reply]
I think you can either just delete the whole `if` statement in the setter or remove `null` coalescing in the getter, right?

**Reply by canton7** [reply]
I'm not going to do either of those: if I remove the coalesce then the consumer needs to check for null, and along with just moving the problem to the consumer I might break someone. If I remove the null check I create a property which doesn't behave like a property, which is bad.

I'm going to make SynchronousDispatcher singleton and expose it.

**Reply by mwdpb** [reply]
I like the singleton idea!

**Reply by canton7** [reply]
See 3959702147f213f412d6229b6c17ff5b411f2cc3

**Reply by mwdpb** [reply]
I see it, I'll check it out.  Thank you!
